File: linkedlists/java linkedlist/src/Linkedlist.java
class Node{
    public int data;
    public Node next;
    public Node(int data) {
        this.data=data;
    }
   


}

public class Linkedlist {
    private Node head;

    public void append(int data){
        if (head == null){
            head = new Node(data);
            return;
        }
        Node temp = head;
        while(temp.next != null){
            temp = temp.next;
        }
        temp.next = new Node(data);
    }

    public void prepend(int data) {
        if (head == null){
            head = new Node(data);
            return;
        }
        Node temp = head;
        head = new Node(data);
        head.next = temp;

    }

    public void delWithValue(int data){
        if(head == null){
            System.out.println("nothing to delete");
            return;
        }
        if (head.data == data){
            head = head.next;
            return;
        }
        Node temp = head;
        while (temp.next != null){
            if(temp.next.data == data){
                temp.next = temp.next.next;
                return;
            }
            temp = temp.next;
        }
        System.out.println("no such value");

    }

    public void traverse(){
        Node temp = head;
        while (temp != null){
            System.out.println(temp.data);
            temp = temp.next;
        }
        
    }

}

File: linkedlists/linkedlist.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int val) {
        data = val;
        next = nullptr;
    }
};

class LinkedList {
private:
    Node* head;

public:
    LinkedList() {
        head = nullptr;
    }

    void insertAtHead(int val) {
        Node* newNode = new Node(val);
        newNode->next = head;
        head = newNode;
    }

    void insertAtTail(int val) {
        Node* newNode = new Node(val);

        if (head == nullptr) {
            head = newNode;
            return;
        }

        Node* current = head;

        while (current->next != nullptr) {
            current = current->next;
        }

        current->next = newNode;
    }

    void deleteAtHead() {
        if (head == nullptr) {
            return;
        }

        Node* temp = head;
        head = head->next;
        delete temp;
    }

    void deleteAtTail() {
        if (head == nullptr) {
            return;
        }

        if (head->next == nullptr) {
            delete head;
            head = nullptr;
            return;
        }

        Node* current = head;
        Node* previous = nullptr;

        while (current->next != nullptr) {
            previous = current;
            current = current->next;
        }

        delete current;
        previous->next = nullptr;
    }

    void printList() {
        if (head == nullptr) {
            cout << "List is empty." << endl;
            return;
        }

        Node* current = head;

        while (current != nullptr) {
            cout << current->data << " ";//
            current = current->next;
        }

        cout << endl;
    }

    void deleteByValue(int val) {
        if (head == nullptr) {
            return;
        }

        if (head->data == val) {
            deleteAtHead();
            return;
        }

        Node* current = head;
        Node* previous = nullptr;

        while (current != nullptr && current->data != val) {
            previous = current;
            current = current->next;
        }

        if (current == nullptr) {
            return;
        }

        previous->next = current->next;
        delete current;
    }
};

int main() {
    LinkedList myList;
    myList.insertAtHead(1);
    myList.insertAtHead(2);
    myList.insertAtTail(3);
    myList.printList(); 
    myList.deleteByValue(3);
    myList.deleteAtTail();
    myList.printList(); 
    return 0;
}